### PR TITLE
Changed grppi:optional to std::experimental::optional  - issue 88

### PR DESCRIPTION
**Pull request by Drioastorga**
I have modified grppi to use internally the experimental optional insted of the grppi's optional. 

Changes:
- Removed grppi optional.
- Modified uses of optional to std::experimental::optional.

### CODE REVIEW

File: include/common/common.h [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 #ifndef GRPPI_COMMON_H
 #define GRPPI_COMMON_H
+//Include standard experimental optional
+#include <experimental/optional>
```
Comment (jdgarciauc3m) [inline include/common/common.h]: I do not think we need to include optional here in any form.

Just include in the places we really use it.

File: tests/farm1.cpp [hunk 2]
```diff
@@ -53,9 +56,9 @@ void farm_example1() {
         [&]() {
             a--; 
             if ( a == 0 ) 
-                return optional<int>(); 
+                return std::experimental::optional<int>();
```
Comment (jdgarciauc3m) [inline tests/farm1.cpp]: Add a:

~~~c++
template <typename T>
using optional = std::experimental::optional<T>;
~~~

before call ing farm.

Comment (jdgarciauc3m) [inline tests/farm1.cpp]: Declare the lambda as

~~~c++
[&a]() -> optional<int> {
  a--;
  if (a==0} return {}
  else return a;
}
~~~

File: unit_tests/map.cpp [hunk 3]
```diff
@@ -18,6 +18,8 @@
 * See COPYRIGHT.txt for copyright notices and details.
 */
 #include <atomic>
+#include <experimental/optional>
```
Comment (jdgarciauc3m) [inline unit_tests/map.cpp]: Why do we need to include optiona here?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
General comments:

1. Simplify code by typedefs (with using syntax) in places like implementations of farm.

2. Be explicit about lambdas by stating return type. In this way we avoid to write verbose code.

File: include/common/common.h (hunk 1 above)
Comment (Drioastorga) [reply include/common/common.h]: Removed from here and included only in the patterns that really need it.

File: tests/farm1.cpp (hunk 2 above)
Comment (Drioastorga) [reply tests/farm1.cpp]: Modified in all tests and unit tests.

File: unit_tests/map.cpp (hunk 3 above)
Comment (Drioastorga) [reply unit_tests/map.cpp]: Removed as it is no tnecessary.

**Comment by Drioastorga** [comment]
I have addresed all the comments but i did not change the use of fully qualified names in the pattern implementations, only in the tests. 

The main reason is that a template can not be defined inside a block, so it has to be declared out of the funcion. Therefore only one declaration of the typedef is allowed, otherwise it will be ambiguous. Even if we define it only once in the grppi code, if the user also define the same it will be ambiguous too.

File: include/native/pipeline.h [hunk 4]
```diff
@@ -19,12 +19,15 @@
 */
 
 #ifndef GRPPI_PIPELINE_THR_H
-#define GRPPI_PIPELINE_THR_H
+#define GRPPI_PIPELINE_THR_Hi
```
Comment (jdgarciauc3m) [inline include/native/pipeline.h]: What is this change?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Address minor comment on include guard.

**Comment by Drioastorga** [comment]
Fixed include guard.

**Review by jdgarciauc3m — APPROVED** [review]
OK

**Comment by jdgarciauc3m** [comment]
Solves issue #88

**Pull request merged**